### PR TITLE
Support fetching chunks

### PR DESCRIPTION
**Pull request by xmakro**
This PR add support for fetching record batches which are served by Snowflake over multiple chunk files. In this case, snowflake sends some initial date via base64, and then a list of additional files via the `resp.data.chunks` field.

In addition, this PR fixes the authentication session header which was broken in a recent commit. This PR also makes the snowflake-rs API thread-safe so clients can issue concurrent requests.

### CODE REVIEW

**Comment by xmakro** [comment]
Description updated, PTAL

**Comment by tbaums** [comment]
@xmakro - Thanks! Taking a look now. Will get back to you shortly!

File: snowflake-api/src/connection.rs [hunk 1]
```diff
@@ -157,4 +157,27 @@ impl Connection {
...
+        url: &str,
+        headers: &HashMap<String, String>,
+    ) -> Result<bytes::Bytes, ConnectionError> {
+        let mut header_map = HeaderMap::new();
+        for (k, v) in headers {
+            header_map.insert(
+                HeaderName::from_bytes(k.as_bytes()).unwrap(),
+                HeaderValue::from_bytes(v.as_bytes()).unwrap(),
+            );
+        }
```
Comment (andrusha) [inline snowflake-api/src/connection.rs:172]: if response `qrmk` field is set and headers are empty, then those would have to be added:

```rust
header_map.insert(
  HeaderName::from_static("x-amz-server-side-encryption-customer-algorithm"),
  HeaderValue::from_static("AES256")
);
let mut qrmk = HeaderValue::from_str(qrmk)?;
qrmk.set_sensitive(true);
header_map.insert(
  HeaderName::from_static("x-amz-server-side-encryption-customer-key"),
  HeaderValue::from_static(qrmk)
);
```

not sure if this is for backwards compatibility in official libs, so could probably just skip it

File: snowflake-api/src/lib.rs [hunk 2]
```diff
@@ -283,22 +281,32 @@ impl SnowflakeApi {
...
         if resp.data.returned == 0 {
-            log::info!("Got response with 0 rows");
-
+            log::debug!("Got response with 0 rows");
             Ok(QueryResult::Empty)
         } else if let Some(json) = resp.data.rowset {
-            log::info!("Got JSON response");
-
+            log::debug!("Got JSON response");
             Ok(QueryResult::Json(json))
```
Comment (andrusha) [inline snowflake-api/src/lib.rs:291]: json response could be chunked too, however I added this more for debugging and status information being passed through that fields. go clients should receive arrow by-default, unless user sets session variable to return json

File: snowflake-api/src/session.rs [hunk 3]
```diff
@@ -40,18 +41,26 @@ pub enum AuthError {
...
 
 #[derive(Debug)]
-pub struct AuthTokens {
-    pub session_token: AuthToken,
-    pub master_token: AuthToken,
+struct AuthTokens {
+    session_token: AuthToken,
+    master_token: AuthToken,
+    /// expected by snowflake api for all requests within session to follow sequence id
+    sequence_id: u64,
```
Comment (andrusha) [inline snowflake-api/src/session.rs:48]: neat!

File: snowflake-api/src/session.rs [hunk 4]
```diff
@@ -207,24 +214,29 @@ impl Session {
...
-            self.auth_tokens = Some(tokens);
+            *auth_tokens = Some(tokens);
         }
-
-        self.auth_tokens.as_ref().ok_or(AuthError::TokenFetchFailed)
+        auth_tokens.as_mut().unwrap().sequence_id += 1;
+        let session_token_auth_header = format!(
+            "Snowflake Token=\"{}\"",
+            &auth_tokens.as_ref().unwrap().session_token.token
+        );
```
Comment (andrusha) [inline snowflake-api/src/session.rs]: there is already method to format token into header string

```
auth_tokens.as_ref().unwrap().session_token. auth_header()
```

**Review by andrusha — APPROVED** [review]

**Comment by andrusha** [comment]
@xmakro could you merge with latest main so it runs CI on your PR?

**Comment by xmakro** [comment]
Thanks for the comments! PTAL, the CI requires approval

**Comment by tbaums** [comment]
sorry! got click happy. @dmzmk said to merge before he went on holiday once you had a chance to review!

**Pull request merged**